File: aidotnet-benchmarks/src/AiDotNetBenchmarks/Controllers/BenchmarkController.cs
using AiDotNetBenchmarks.Benchmarks;
using Microsoft.AspNetCore.Mvc;

namespace AiDotNetBenchmarks.Controllers;

/// <summary>
/// Runs the four-model AiDotNet benchmark (MLP / CNN / LSTM / Transformer) and
/// returns the JSON report — the controller-shaped equivalent of the PyTorch
/// side's <c>POST /api/Benchmark/Models</c> (and of the CLI scaffold this repo
/// briefly carried). Same workload knobs as the PyTorch CLI defaults: 3 epochs ×
/// 20 batches × bs64 training; 100 steady-state inference iterations after 10
/// warmups at batch sizes 1/8/32/128.
/// </summary>
/// <remarks>
/// A full four-model run takes minutes — use a generous client timeout. The run
/// executes in-process: for the publication-grade methodology (fresh process per
/// round, interleaved against PyTorch, min-of-N p95 per shape) restart the host
/// between rounds; see Reporting/findings.md.
/// </remarks>
[ApiController]
[Route("api/[controller]")]
public sealed class BenchmarkController : ControllerBase
{
    // One benchmark at a time: the runner pins the global engine, suppresses GC
    // around steady-state loops, and saturates the CPU — two concurrent runs
    // would corrupt each other's measurements (and a no-GC region cannot nest).
    private static readonly SemaphoreSlim RunGate = new(1, 1);

    [HttpGet("Test")]
    public ActionResult<string> Test() => "BenchmarkController is ready. POST api/Benchmark/Models to run.";

    /// <summary>
    /// Run the benchmark for the requested model families and return the report.
    /// </summary>
    /// <param name="models">Comma-separated subset of mlp,cnn,lstm,transformer,mlp-fused.</param>
    /// <param name="epochs">Training epochs per model.</param>
    /// <param name="trainBatches">Train batches per epoch.</param>
    /// <param name="batchSize">Training batch size.</param>
    /// <param name="inferenceIterations">Steady-state inference iterations per batch size.</param>
    /// <param name="warmupIterations">Warmup iterations before the steady-state window.</param>
    /// <param name="seed">Deterministic synthetic-data seed.</param>
    [HttpPost("Models")]
    public async Task<ActionResult<BenchmarkReport>> Models(
        [FromQuery] string models = "mlp,cnn,lstm,transformer",
        [FromQuery] int epochs = 3,
        [FromQuery] int trainBatches = 20,
        [FromQuery] int batchSize = 64,
        [FromQuery] int inferenceIterations = 100,
        [FromQuery] int warmupIterations = 10,
        [FromQuery] int seed = 1234)
    {
        var modelNames = models.Split(',', StringSplitOptions.RemoveEmptyEntries | StringSplitOptions.TrimEntries);
        if (modelNames.Length == 0)
            return BadRequest(new { error = "Provide at least one model: mlp, cnn, lstm, transformer, mlp-fused." });
        if (epochs < 1 || trainBatches < 1 || batchSize < 1 || inferenceIterations < 1 || warmupIterations < 0)
            return BadRequest(new { error = "Workload parameters must be positive (warmupIterations may be 0)." });

        if (!await RunGate.WaitAsync(TimeSpan.Zero, HttpContext.RequestAborted))
        {
            return StatusCode(StatusCodes.Status409Conflict, new
            {
                error = "A benchmark run is already in progress. Concurrent runs would contend for CPU and corrupt both measurements."
            });
        }

        try
        {
            var options = new BenchmarkOptions(
                modelNames, epochs, trainBatches, batchSize, inferenceIterations, warmupIterations, seed);
            // The runner is CPU-bound for minutes; keep it off the request thread
            // so Kestrel's loop stays responsive.
            var report = await Task.Run(() => new BenchmarkRunner(options).Run(), HttpContext.RequestAborted);
            return Ok(report);
        }
        catch (ArgumentException ex)
        {
            return BadRequest(new { error = ex.Message });
        }
        finally
        {
            RunGate.Release();
        }
    }
}

File: pytorch-benchmarks/src/pytorch_benchmarks/controllers/benchmark_controller.py
"""Controller route for the four-model PyTorch benchmark.

Mirrors the AiDotNet side's ``POST /api/Benchmark/Models`` so the
``BothController`` fan-out (and any developer with two running hosts) can
trigger both frameworks' benchmarks over HTTP and compare the JSON reports —
the same controller-shaped workflow as the regression endpoints.

The implementation delegates to the exact same ``run(...)`` used by the
``python -m pytorch_benchmarks`` CLI, so endpoint mode and CLI mode measure
identical workloads.
"""

from __future__ import annotations

import argparse
import threading

from fastapi import APIRouter
from fastapi.responses import JSONResponse

from pytorch_benchmarks.__main__ import run

router = APIRouter(prefix="/api/Benchmark", tags=["Benchmark"])

# One benchmark at a time: a run saturates the CPU for minutes; two concurrent
# runs would each be measuring the other's contention. Mirrors the AiDotNet
# BenchmarkController's run gate.
_run_gate = threading.Lock()


@router.get("/Test")
def test() -> str:
    return "Benchmark controller is ready. POST api/Benchmark/Models to run."


@router.post("/Models")
def models(
    models: str = "mlp,cnn,lstm,transformer",
    epochs: int = 3,
    trainBatches: int = 20,
    batchSize: int = 64,
    inferenceIterations: int = 100,
    warmupIterations: int = 10,
    seed: int = 1234,
    device: str = "cpu",
    threads: int = 8,
) -> JSONResponse:
    """Run the benchmark for the requested model families and return the report.

    Query parameter names match the AiDotNet ``BenchmarkController`` so the
    ``/api/Both/Benchmark`` fan-out can forward one query string to both hosts.
    ``device`` defaults to ``cpu`` and ``threads`` to 8 — the fair-comparison
    defaults (the AiDotNet side pins its CPU engine and
    ``AIDOTNET_BLAS_THREADS=8``).
    """
    names = [item.strip() for item in models.split(",") if item.strip()]
    if not names:
        return JSONResponse(
            status_code=400,
            content={"error": "Provide at least one model: mlp, cnn, lstm, transformer."},
        )
    if epochs < 1 or trainBatches < 1 or batchSize < 1 or inferenceIterations < 1 or warmupIterations < 0:
        return JSONResponse(
            status_code=400,
            content={"error": "Workload parameters must be positive (warmupIterations may be 0)."},
        )

    if not _run_gate.acquire(blocking=False):
        return JSONResponse(
            status_code=409,
            content={
                "error": "A benchmark run is already in progress. Concurrent runs would "
                         "contend for CPU and corrupt both measurements."
            },
        )
    try:
        args = argparse.Namespace(
            models=",".join(names),
            device=device,
            epochs=epochs,
            train_batches=trainBatches,
            batch_size=batchSize,
            inference_iterations=inferenceIterations,
            warmup_iterations=warmupIterations,
            seed=seed,
            threads=threads,
        )
        try:
            report = run(args)
        except ValueError as exc:  # unknown model name from make_model
            return JSONResponse(status_code=400, content={"error": str(exc)})
        return JSONResponse(status_code=200, content=report)
    finally:
        _run_gate.release()
